gap/congruences/congsemigraph.gi

```diff
@@ -0,0 +1,271 @@
+############################################################################
+##
+##  congsemigraph.gi
+##  Copyright (C) 2021                       Marina Anagnostopoulou-Merkouri
+##                                                            James Mitchell
+##
+##  Licensing information can be found in the README file of this package.
+##
+############################################################################
+
+BindGlobal("SEMIGROUPS_IsHereditarySubset",
+function(S, H)
+  local out, h, v, D;
+  D := GraphOfGraphInverseSemigroup(S);
+  if IsMultiDigraph(D) then
+    ErrorNoReturn(D, " must not have multiple edges");
+  fi;
+  out := OutNeighbours(D);
+  for h in H do
+    for v in out[h] do
+      if not (v in H) then
+        return false;
+      fi;
+    od;
+  od;
+  return true;
+end);
+
+BindGlobal("SEMIGROUPS_IsValidWSet",
+function(S, H, W)
+  local w, c, v, out, D;
+  if not SEMIGROUPS_IsHereditarySubset(S, H) then
+    ErrorNoReturn(H, "must be a valid hereditary subset");
+  fi;
+  D := GraphOfGraphInverseSemigroup(S);
+  if IsMultiDigraph(D) then
+    ErrorNoReturn(D, "must not have multiple edges");
+  fi;
+  out := OutNeighbours(D);
+  for w in W do
+    c := 0;
+    for v in out[w] do
+      if v in H then
+        c := c + 1;
+      fi;
+    od;
+    if c <> Length(out[w]) - 1 then
+      return false;
+    fi;
+  od;
+  return true;
+end);
+
+BindGlobal("SEMIGROUPS_ValidateWangPair",
+function(S, H, W)
+  local v, D;
+  D := GraphOfGraphInverseSemigroup(S);
+  for v in Union(H, W) do
+    if not (v in DigraphVertices(D)) then
+      ErrorNoReturn("the elements of ", H, " and ",
+                    W, " are not all vertices of ", D);
+    fi;
+  od;
+  if not SEMIGROUPS_IsHereditarySubset(S, H) then
+    ErrorNoReturn(H, " is not a valid hereditary subset");
+  fi;
```

Comment:
```suggestion
  elif not SEMIGROUPS_IsValidWSet(S, H, W) then
```

To save one line

Reply:
I suggest making `SEMIGROUPS_IsHereditarySubset` and `SEMIGROUPS_IsValidWSet` always return `true` or `false` (no errors), and writing the errors in this function instead. This would be cleaner, and remove some duplicate code (for `IsMultiDigraph` for example).